frontend/app/routes/auth.$.tsx

```diff
@@ -177,3 +187,13 @@ function handleMockAuthorizeRequest({ params, request }: LoaderFunctionArgs) {
   log.debug('Mock login successful; redirecting to [%s]', redirectUri.toString());
   return redirect(redirectUri.toString());
 }
+/**
+ * Replaces all variables in a string with their corresponding values.
+ * Variables are denoted as `{variableName}`.
+ */
+function replaceVariables(str: string, variables: Record<string, string>) {
```

Comment:
Move it in `string-utils.ts` file and write unit tests.

Reply:
Done. Please re-review.